src/main.rs

```diff
@@ -237,26 +237,33 @@ async fn entry() -> Result<()> {
         .map_err(BinstallError::from)
         .wrap_err("Creating a temporary directory failed.")?;
 
-    info!("Installing package: '{}'", opts.name);
+    info!("Installing package: '{}'", opts.crate_name);
+
+    let version = match (&opts.crate_name.version, &opts.version) {
+        (Some(version), None) => version.to_string(),
+        (None, Some(version)) => version.to_string(),
+        (Some(_), Some(_)) => Err(BinstallError::DuplicateVersionReq)?,
+        (None, None) => "*".to_string(),
+    };
 
     // Fetch crate via crates.io, git, or use a local manifest path
     // TODO: work out which of these to do based on `opts.name`
     // TODO: support git-based fetches (whole repo name rather than just crate name)
     let manifest = match opts.manifest_path.clone() {
         Some(manifest_path) => load_manifest_path(manifest_path.join("Cargo.toml"))?,
-        None => fetch_crate_cratesio(&client, &opts.name, &opts.version).await?,
+        None => fetch_crate_cratesio(&client, &opts.crate_name.name, &version).await?,
     };
 
     let package = manifest.package.unwrap();
 
-    let is_plain_version = semver::Version::from_str(&opts.version).is_ok();
-    if is_plain_version && package.version != opts.version {
+    let is_plain_version = semver::Version::from_str(&version).is_ok();
+    if is_plain_version && package.version != version {
```

Comment:
I wonder if now would be a good time to change that behaviour, such that `1.2.3` parses as `=1.2.3` instead of `^1.2.3`. That matches more closely what cargo install does, and would get rid of this ambiguity (and the warning).

Reply:
Done!